Add parse bin int, fixes #8018

Vindaar opened the pull request:
I agree with #8018 that it would be nice to offer such a proc. So I just took the implementation of parseutil `parseBin` and removed the parts related to the return value.

I also added some simple test cases for all three parse*Int procs. 

One question: when I saw the implementation for `parseHexInt` and `parseOctInt`. Is there a specific reason, why they don't simply wrap `parseHex` and `parseOct`? For efficiency reasons?

Araq commented on lib/pure/strutils.nim:
```diff
@@ -906,7 +906,24 @@ proc parseHexInt*(s: string): int {.noSideEffect, procvar,
+    of '0'..'1':
+      result = result shl 1 or (ord(s[i]) - ord('0'))
+    else: raise newException(ValueError, "invalid binary integer: " & s)
+    inc(i)
```
There is a check missing here testing that we've processed at least one digit.

Vindaar replied:
I'm not following. Do you mean regarding the condition to raise the exception?

Araq replied:
``0b`` or even the empty string would parse successfully with your solution, but shouldn't.

Vindaar replied:
Oh, I thought that was intended behavior. `parseHexInt` and `parseOctInt` are working the same way. Should that be changed for all 3 then?

dom96 replied:
> Should that be changed for all 3 then?

I'd say: yes.

Araq commented:
Looks very good otherwise.

> Is there a specific reason, why they don't simply wrap parseHex and parseOct? For efficiency reasons?

Legacy.

Vindaar commented:
> > Is there a specific reason, why they don't simply wrap parseHex and parseOct? For efficiency reasons?

> Legacy.

Should I create a PR to do that?

Araq commented:
> Should I create a PR to do that?

I you feel strongly about it.

data-man commented:
I already proposed this (in the chat).
Attempts №2:

- remove duplicated code from ```strutils.parse*Int``` - use ```parseutils.parse*``` instead. Like in the ```strutils.parseInt```
- add ```maxLen``` to ```parseOct``` and ```parseBin```

Ideally - to add ```parse(bin, oct, hex)[T]: T``` for supports int64 on x32.

Vindaar commented:
So I just went ahead and replaced the implementations for the procs by the ones in `parseutils`. I thought that was cleaner than copy pasting the `foundDigit` pieces from there.

Also replaced a couple of `var` with `let` in other `parse` procs, which also only wrap `parseLen`.

Finally, went ahead and followed @data-man's suggestion to include the optional `maxLen` argument to `parseOct` and `parseBin` in parseutils.

Varriount commented on tests/stdlib/tstrutil.nim:
```diff
@@ -7,6 +7,11 @@ discard """
 import
   strutils
 
+template rejectParse(e) =
```
This just ignores the expression if it throws a value error. Shouldn't it fail if a value error isn't raised?

Vindaar replied:
Haha, indeed. Thanks! I even added a test case, that's obviously not going to fail (checking for "0bFF" with `parseHexInt`).

Varriount commented on lib/pure/parseutils.nim:
```diff
@@ -87,14 +91,18 @@ proc parseOct*(s: string, number: var int, start = 0): int  {.
-  if i+1 < s.len and s[i] == '0' and (s[i+1] == 'b' or s[i+1] == 'B'): inc(i, 2)
-  while i < s.len:
+  let last = if maxLen == 0: s.len else: i+maxLen  
+  if i+1 < last and s[i] == '0' and (s[i+1] == 'b' or s[i+1] == 'B'): inc(i, 2)
```
You can use a set here to check if `s[i+1]` is b or B.

Vindaar replied:
Good point!

Varriount commented on lib/pure/parseutils.nim:
```diff
@@ -87,14 +91,18 @@ proc parseOct*(s: string, number: var int, start = 0): int  {.
   var foundDigit = false
-  if i+1 < s.len and s[i] == '0' and (s[i+1] == 'b' or s[i+1] == 'B'): inc(i, 2)
-  while i < s.len:
+  let last = if maxLen == 0: s.len else: i+maxLen  
```
What if the passed in maximum length is greater than the string length?

Vindaar replied:
That's a very good point. I used the same code as for `parseHex`, without checking it for any bugs (there are no test cases for parseutils, as far as I can tel...). I'll go ahead and fix that for all three.

Vindaar commented:
@Varriount Should be all fixed now. :)

Varriount commented on tests/stdlib/tstrutil.nim:
```diff
@@ -7,6 +7,14 @@ discard """
 import
   strutils
 
+import macros
```
Why is `macros` imported?

dom96 replied:
``astToStr`` I guess

Vindaar replied:
Yup. Or is there a simpler way to echo the code handed to the template?

Araq replied:
``astToStr`` is in system!

Vindaar replied:
Ooops, sorry! Please remove that import... :/

dom96 commented on changelog.md:
```diff
@@ -99,7 +101,7 @@
 - Added the parameter ``val`` for the ``CritBitTree[T].incl`` proc.
 - The proc ``tgamma`` was renamed to ``gamma``. ``tgamma`` is deprecated.
-
+- The procs ``parseHexInt`` and ``parseOctInt`` now fail on empty strings and strings containing only valid prefixes, e.g. "0x" for hex integers.
```
This should be under "breaking changes to the stdlib"

Vindaar replied:
Ok, I wasn't sure!

dom96 commented on lib/pure/parseutils.nim:
````diff
@@ -47,12 +47,14 @@ proc parseHex*(s: string, number: var int, start = 0; maxLen = 0): int {.
   ##
-  ## If 'maxLen==0' the length of the hexadecimal number has no
-  ## upper bound. Not more than ```maxLen`` characters are parsed.
+  ## If `maxLen == 0` the length of the hexadecimal number has no upper bound.
````
Double `

Vindaar replied:
Done.
